### PR TITLE
Incremental resource updates for NgCompiler, Language Service

### PR DESCRIPTION
**Pull request by alxhub**

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/annotations/src/component.ts [hunk 1]
```diff
@@ -978,4 +1002,24 @@ export interface ParsedComponentTemplate extends ParsedTemplate {
...
+  declaration: TemplateDeclaration;
 }
+
+interface CommonTemplateDeclaration {
+  preserveWhitespaces: boolean;
+  interpolationConfig: InterpolationConfig;
+  templateUrl: string;
+  resolvedTemplateUrl: string;
+  sourceMapUrl: string;
+}
```
Comment (atscott) [inline packages/compiler-cli/src/ngtsc/annotations/src/component.ts:1052]:
```suggestion
}

```
nit

File: packages/compiler-cli/src/ngtsc/annotations/src/component.ts [hunk 2]
```diff
@@ -978,4 +1002,24 @@ export interface ParsedComponentTemplate extends ParsedTemplate {
 
 export interface ParsedTemplateWithSource extends ParsedComponentTemplate {
   sourceMapping: TemplateSourceMapping;
+  declaration: TemplateDeclaration;
 }
+
+interface CommonTemplateDeclaration {
```
Comment (atscott) [inline packages/compiler-cli/src/ngtsc/annotations/src/component.ts:1046]: nit: maybe some short jsdoc about these. Specifically about how the declaration is unparsed and why we do it in two steps (basically a shorter version of the commit comment)

File: packages/compiler-cli/src/ngtsc/core/src/compiler.ts [hunk 3]
```diff
@@ -306,6 +327,36 @@ export class NgCompiler {
     this.ignoreForEmit = this.adapter.ignoreForEmit;
   }
 
+  private updateWithChangedResources(changedResources: Set<string>): void {
+    if (this.compilation === null) {
+      // Analysis hasn't happened yet, so no update is necessary - any changes to resources will be
+      // captured by the
```
Comment (atscott) [inline packages/compiler-cli/src/ngtsc/core/src/compiler.ts:333]: captured by the...? :)

File: packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts [hunk 4]
```diff
@@ -318,6 +318,26 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
     return engine.getExpressionCompletionLocation(ast);
   }
 
+  invalidateClass(clazz: ts.ClassDeclaration): void {
+    this.completionCache.delete(clazz);
+    this.symbolBuilderCache.delete(clazz);
+    this.scopeCache.delete(clazz);
+    this.elementTagCache.delete(clazz);
+
+
```
Comment (atscott) [inline packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:327]: nit: extra new line

File: packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts [hunk 5]
```diff
@@ -318,6 +318,26 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
     return engine.getExpressionCompletionLocation(ast);
   }
 
+  invalidateClass(clazz: ts.ClassDeclaration): void {
```
Comment (atscott) [inline packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:321]: nit: short jsdoc

**Review by atscott — APPROVED** [review]
Awesome work! This is a really exciting improvement to the incremental compilation!

File: packages/compiler-cli/src/ngtsc/tsc_plugin.ts [hunk 6]
```diff
@@ -101,10 +101,20 @@ export class NgTscPlugin implements TscPlugin {
...
+    let ticket: CompilationTicket;
+    if (oldProgram === undefined || oldDriver === null) {
+      ticket = freshCompilationTicket(
+          program, this.options, strategy, typeCheckStrategy,
+          /* enableTemplateTypeChecker */ false, /* usePoisonedData */ false);
+    } else {
+      strategy.toNextBuildStrategy().getIncrementalDriver(oldProgram);
+      ticket = incrementalFromDriverTicket(
+          oldProgram, oldDriver, program, this.options, strategy, typeCheckStrategy,
+          this.host.getModifiedResourceFiles?.() ?? new Set<string>(), false, false);
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/tsc_plugin.ts:115]: Hahaha. A bit too clever :)

**Comment by sod** [comment]
the new ngtsc base ivy webpack plugin is stil 15x slower for us. Thx for keeping around the old one I guess? :)

Seems to stil compile everything even when changing a html file. Project: 206 components, angular 11.1.1, profile of 2nd change in a component html file: [CPU-20210128T155138.cpuprofile.zip](https://github.com/angular/angular/files/5888106/CPU-20210128T155138.cpuprofile.zip)

ivy plugin: ~20 sec., old plugin ~1.5 sec.

**Comment by JoostK** [comment]
@sod The Ivy Webpack plugin is unrelated to this change; please open an issue in the CLI repo where the Webpack integration is maintained.

**Comment by bbarry** [comment]
Is there any general documentation about how this feature helps developers (like what the various incremental steps are by default available in an `ng build` scenario)?

Reading the code here it looks like there is a reusable compilation unit here for the component when the template changes? Does that only apply when you have a `--watch` style scenario (`ng build --watch` or `ng serve` for example)?

**Pull request closed without merging**